common/peerprovider/plugin.go

```diff
@@ -0,0 +1,87 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package peerprovider
+
+import (
+	"fmt"
+
+	"go.uber.org/yarpc/transport/tchannel"
+
+	"github.com/uber/cadence/common/config"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/membership"
+	"github.com/uber/cadence/common/syncmap"
+)
+
+// Container is passed to peer provider plugin
+type Container struct {
+	Service string
+	// Channel is required by ringpop
+	Channel tchannel.Channel
+	Logger  log.Logger
+	Portmap membership.PortMap
+}
+
+type constructorFn func(cfg *config.YamlNode, container Container) (membership.PeerProvider, error)
+
+var plugins = syncmap.New[string, plugin]()
+
+type plugin struct {
+	fn        constructorFn
+	configKey string
+}
+
+type Provider struct {
+	config    config.PeerProvider
+	container Container
+}
+
+func New(config config.PeerProvider, container Container) *Provider {
+	return &Provider{
+		config:    config,
+		container: container,
+	}
+}
+
+func Register(configKey string, constructor constructorFn) error {
+	inserted := plugins.Put(configKey, plugin{
+		fn:        constructor,
+		configKey: configKey,
+	})
+
+	if !inserted {
+		return fmt.Errorf("peer provider %q is already registered", configKey)
+	}
+
+	return nil
+}
+
+func (p *Provider) Provider() (membership.PeerProvider, error) {
+	for configKey, cfg := range p.config {
```

Comment:
do you need multiple provider support? If so the config should have a place to define these list of providers and another place to pick which one is active. Returning the first registered one might cause unintended provider to be picked

Reply:
adding more thoughts here: at least in theory, you might use multiple peer providers at the same time, for migration needs. Will revisit this method to check what it takes to support this scenario.

Reply:
Multiple provider support will increase complexity in membership package highly, what we can do instead, limit provider number to 1, if anyone needs multi-source peer provider solution, you can replace plugin with  multi-source plugin. This would keep internal membership logic unchanged and there will be no requirement to consult multiple peer providers.